Fix #4355: Added refresher exploration redirection in player view

aks681 opened the pull request:
Added a redirection confirmation modal whenever user clicks on a wrong answer that gives the option to the user to redirect to a short refresher exploration and come back later to do the questions.

**Checklist**
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes.
- [x] The linter/Karma presubmit checks have passed.
  - These should run automatically, but if not, you can manually trigger them locally using `python scripts/pre_commit_linter.py` and `bash scripts/run_frontend_tests.sh`.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR follows the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide).
- [x] The PR is assigned to an appropriate reviewer.
  - If you're a new contributor, please ask on [Gitter](https://gitter.im/oppia/oppia-chat) for someone to assign a reviewer.
  - If you're not sure who the appropriate reviewer is, please assign to the issue's "owner" -- see the "talk-to" label on the issue.

aks681 committed:
Added refresher exploration confirmation modal

seanlip commented on core/templates/dev/head/pages/exploration_player/refresher_exploration_confirmation_modal_directive.html:
```diff
@@ -0,0 +1,15 @@
+
+<div class="modal-body">
+  <span class="oppia-learner-refresher-exploration-confirmation-body">
+    t looks like you're having some trouble answering the question.
```
It

aks681 replied:
Done

seanlip commented on core/templates/dev/head/css/oppia.css:
```diff
@@ -122,6 +122,10 @@ md-input-group.md-default-theme label
   word-wrap: break-word;
 }
 
+.oppia-learner-refresher-exploration-confirmation-body {
```
Put these new classes in the corresponding modal directive -- easier to keep track of them that way. (At some point we should break oppia.css up and put stuff into the relevant directives, otherwise everything gets loaded every time.)

aks681 replied:
Done

seanlip commented on core/templates/dev/head/pages/exploration_player/ConversationSkinDirective.js:
```diff
@@ -565,6 +567,8 @@ oppia.directive('conversationSkin', [
                         ExplorationPlayerService.getRandomSuffix());
                     }
+                    RefresherExplorationConfirmationModalService.
+                      displayRedirectConfirmation(refresherExplorationId);
```
displayRedirectConfirmationModal

aks681 replied:
Done

seanlip commented on core/templates/dev/head/pages/exploration_player/ConversationSkinDirective.js:
```diff
@@ -565,6 +567,8 @@ oppia.directive('conversationSkin', [
                           newStateName, _nextFocusLabel) +
                         ExplorationPlayerService.getRandomSuffix());
                     }
+                    RefresherExplorationConfirmationModalService.
```
Before this, you'll need to check whether the exploration that's being redirected to exists and is public -- perhaps by making a call to library.ExplorationSummariesHandler in the backend. This is an asynchronous check, but it should be OK to display the modal only when the response arrives.

(The status of the exploration may have changed since it was added to the current exploration as a refresher one.)

aks681 replied:
Done, should this be extended to editor also, when author gives the id?

seanlip commented on core/templates/dev/head/pages/exploration_player/RefresherExplorationConfirmationModalService.js:
```diff
@@ -0,0 +1,51 @@
+            function($scope, $uibModalInstance, $window) {
+              $scope.confirmRedirect = function() {
+                var url = '/explore/' + refresherExplorationId;
+                url = UrlService.addField(
```
Nice use of your new library! :)

aks681 replied:
Thanks!

seanlip reviewed (commented):
Thanks @aks681! Looks generally good, I had a few small comments.

Could you also please do an e2e test for this (probably better to add a new test altogether rather than modify the existing one)? Basically, create an exp and set a redirect ID on it, then play it and make sure the redirect behaviour works (the modal comes up, etc., and cancelling does nothing, and resubmitting an answer and confirming leads to a different exp).

Thanks!

aks681 committed:
made review changes

aks681 commented:
Will do the e2e tests now.
Also, one thing, when user clicks yes on redirect confirmation, the browser by default shows a 'Are you sure you want to exit this page? Work won't be saved' dialog. Would that be a problem?

seanlip commented:
Oh, yes, good catch. I think we add that check somewhere else (on an onBeforeUnload handler?), actually, so we should be able to disable it in this case.

aks681 committed:
disabled browser warning when redirecting

aks681 committed:
fixed lint errors

aks681 committed:
minor change

aks681 commented:
Yes, I added it in the eventListener on 'beforeunload' in ConverstionSkinDirective.js, 
@seanlip, can you check if that's fine?

seanlip commented on core/templates/dev/head/pages/exploration_player/ConversationSkinDirective.js:
```diff
@@ -565,6 +569,25 @@ oppia.directive('conversationSkin', [
+                          displayRedirectConfirmationModal(
+                            refresherExplorationId,
+                            function(
+                              redirectConfirmed) {
```
Can move this to previous line.

Also, do you even need an arg to be passed? Couldn't you just have the callback trigger in the success case?

aks681 replied:
Done, yes, I removed the arg.

seanlip reviewed (commented):
Yep, the beforeUnload approach seems fine. Thanks! Just one minor comment, otherwise I think it's just the e2e test that's left.

seanlip commented:
Note that there are some Travis failures on existing tests which might be worth looking into. This is a bit odd since I didn't think any of the existing tests had this functionality, so you might want to check them out and make sure the vanilla case passes. In particular, existing explorations shouldn't be making calls to the explorationsummarieshandler.

aks681 committed:
made review changes

aks681 commented:
Yes, I added a check for existence of refresher exploration id before making the calls, would that resolve the problem?

seanlip commented:
Yep! I think so.

aks681 commented:
@seanlip Its still failing in some tests that my changes have no effect on.

seanlip commented:
I'm looking at the error message. I see:

```
[chrome #01-1]   Message:
[chrome #01-1]     Expected [ Entry({ level: SEVERE, message: 'http://localhost:9001/third_party/static/angularjs-1.5.8/angular.min.js 117:9 "Error: [$injector:unpr] http://errors.angularjs.org/1.5.8/$injector/unpr?p0=RefresherExplorationConfirmationModalServiceProvider%20%3C-%20RefresherExplorationConfirmationModalService\n    at 1515061973857, type: '' }) ] to equal [  ].
```

which does seem to indicate that some dependency is not imported correctly?

Try running through the test manually on your computer (i.e. look at the protractor test file and manually simulate the same actions as a user), keeping your eye on the browser console while doing so. That may help with debugging.

aks681 commented:
Ok, will try that

aks681 committed:
added e2e test

aks681 committed:
removed debug log

aks681 commented:
Added the e2e test, also I think the error was that I didn't import file in exploration_editor.html as in preview tab of editor also, the service was called. After I added that, the error was gone when I ran locally.

seanlip reviewed (approved):
Looks good. Thanks, @aks681!

aks681 committed:
merged upstream/develop

aks681 commented:
Thank you!

aks681 committed:
fixed error due to merge conflict
